src/main/java/com/iexec/core/security/JwtTokenProvider.java

```diff
@@ -42,13 +42,25 @@ public JwtTokenProvider(ChallengeService challengeService) {
         this.secretKey = Base64.getEncoder().encodeToString(seed);
     }
 
+    /**
+     * Creates a signed JWT with expiration date for a given ethereum address.
+     * <p>
+     * The token is cached until it expires.
+     * @param walletAddress worker address for which the token is created
+     * @return A signed JWT for a given ethereum address
+     */
     public String createToken(String walletAddress) {
```

Comment:
Shouldn't we check the expiration of the token there as well?

Reply:
The `createToken` is only called from `WorkerController#login` which should almost only happen when a worker receives an `UNAUTHORIZED` HTTP status.
A worker receives such a status when a token is not valid (not signed, expired or bad claims).
This is true for the `/replicates/available`, `/replicates/interrupted`, `/replicates/{chainTaskId}/status`, `/workers/ping`, `/workers/register`, `/workers/computing` endpoints which should be used more often than the `/worker/login` one.

I think it would be better to keep the current algorithm here:
- The worker provides a token on an aforementioned endpoint
- When the token is expired
  - it is purged from its cache (as well as the challenge)
  - the worker receives HTTP status UNAUTHORIZED
  - the worker launches the login procedure in `LoginService` (create new challenge, create new token)
- The token is valid on next REST call requiring it

With this flow, almost no call to `/workers/login` should be made with an expired token still existing in the cache.

To go your way would require to modify `createToken`:
- to check if a value/token exists for the provided address
- if it exists, to check if it is valid
  - when valid, return the token
  - if invalid, challenge and token will be purged from maps
  - call to `/worker/login` will fail in worker

It seems to be a more complex implementation which only allow to gain one HTTP request.
I am not sure the gain is really worth the investment.
What do you think ?

Reply:
It doesn't seem it would need a lot of investment. I guess something like that would work:
```java
public String createToken(String walletAddress) {
    final String token = jwTokensMap.get(walletAddress);
    if (!StringUtils.isEmpty(token) && isValidToken(token)) {
        // token is valid, reusing it
        return token;
    }

    // token is invalid, renewing it
    challengeService.removeChallenge(walletAddress);
    return jwTokensMap.compute(walletAddress, (address, oldToken) -> {
        Date now = new Date();
        return Jwts.builder()
                .setAudience(address)
                .setIssuedAt(now)
                .setExpiration(new Date(now.getTime() + TOKEN_VALIDITY_DURATION))
                .setSubject(challengeService.getChallenge(address))
                .signWith(SignatureAlgorithm.HS256, secretKey)
                .compact();
    });
}
```

In fact, I usually find it better to avoid communications (HTTP requests) when they are not required.

However, if we decide to keep your version, we should at least fix the JavaDoc. It currently says the following:
> The token is cached until it expires.
This is not true, as this method can return an expired token that will be purged only on next authenticated call.

Reply:
see b56c839